<div style="text-align:left;font-size:2em"><span style="font-weight:bolder;font-size:1.25em">SP2273 | Learning Portfolio</span><br><br><span style="font-weight:bold;color:darkred">Loops (Good) Exercises</span></div>

In [7]:
#exercise 1: making an odd list to generate squares of the odd integers from 0 to 9
list = []
for n in range(10):
    if n % 2 == 0:
        continue
    list.append(n**2)
print(list)

[1, 9, 25, 49, 81]


In [6]:
#exercise 2: doing the same thing using list comprehension meaning output at the front

[n**2 for n in range(10) if n % 2 == 1]

[1, 9, 25, 49, 81]

In [17]:
#exercise 3 timeit
%timeit list = []
for n in range(10):
    if n % 2 == 0:
        continue
    list.append(n**2)
print(list)

58.7 ns ± 6.45 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
[1, 9, 25, 49, 81, 1, 9, 25, 49, 81, 1, 9, 25, 49, 81]


In [18]:
%timeit [n**2 for n in range(10) if n % 2 == 1]

2.39 µs ± 281 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


Ans: the former solution with the for loop is faster at 58.7 ns as compared to the one using list comprehension, which took ard 2.39 us

In [25]:
#exercise 4: math decay half life - figuring out how many years it will take
x = 0
mass = 100
while mass >= 50:
    x += 1
    mass = .95*mass
    if mass >= 50:
        print(f"at {x} years, the sample weighs {mass} units")

#ans: takes 13 years

at 1 years, the sample weighs 95.0 units
at 2 years, the sample weighs 90.25 units
at 3 years, the sample weighs 85.7375 units
at 4 years, the sample weighs 81.45062499999999 units
at 5 years, the sample weighs 77.37809374999999 units
at 6 years, the sample weighs 73.50918906249998 units
at 7 years, the sample weighs 69.83372960937498 units
at 8 years, the sample weighs 66.34204312890623 units
at 9 years, the sample weighs 63.02494097246091 units
at 10 years, the sample weighs 59.87369392383786 units
at 11 years, the sample weighs 56.880009227645964 units
at 12 years, the sample weighs 54.03600876626366 units
at 13 years, the sample weighs 51.33420832795048 units


In [35]:
#exercise 5: atmospheric co2 lvls. task: identify years with > 3ppm increase, print out years along w corresponding change
co2_data = [
    (2000, 369.55), (2001, 371.14), (2002, 373.28), 
    (2003, 375.80), (2004, 377.52), (2005, 379.80), 
    (2006, 381.90), (2007, 383.79), (2008, 385.60), 
    (2009, 387.43), (2010, 389.90), (2011, 391.65), 
    (2012, 393.85), (2013, 396.52), (2014, 398.65),
    (2015, 400.83), (2016, 404.24), (2017, 406.55), 
    (2018, 408.52), (2019, 411.44), (2020, 414.24)
]

conc_prev = 0
for year, conc in co2_data:
    difference = conc - conc_prev
    print(year, difference)
    conc_prev = conc #setting the new "benchmark" as that year before new loop begins

2000 369.55
2001 1.589999999999975
2002 2.1399999999999864
2003 2.5200000000000387
2004 1.7199999999999704
2005 2.2800000000000296
2006 2.099999999999966
2007 1.8900000000000432
2008 1.8100000000000023
2009 1.829999999999984
2010 2.4699999999999704
2011 1.75
2012 2.2000000000000455
2013 2.669999999999959
2014 2.1299999999999955
2015 2.180000000000007
2016 3.410000000000025
2017 2.3100000000000023
2018 1.9699999999999704
2019 2.920000000000016
2020 2.8000000000000114


In [41]:
#part 2: filtering out the years
conc_prev = 0
for year, conc in co2_data:
    difference = conc - conc_prev
    conc_prev = conc #setting the new "benchmark" as that year before new loop begins
    if difference < 3 or year == 2000:
        continue
    print(f"{year} showed an increase of {difference} ppm as compared to the prev year")

2016 showed an increase of 3.410000000000025 ppm as compared to the prev year
